### ISSUE TITLE
Question: transitioning out when component no longer rendered

### ISSUE THREAD

**Issue by jacobweber** [reactjs/react-transition-group#431]
Sorry if this isn't the right forum. I'm new to this library, and I have a basic question about transitioning out.

I have a parent component that conditionally renders several child components wrapped in transitions. I'd like those child components to transition in when they're rendered (when the condition becomes true), and out when they're removed (when it becomes false). So I'm just setting `in={true}` for all of the children. With this approach, the in-transition works, but the out-transition doesn't; the components are just removed when their conditions become false, with no animation.

Is this a fundamentally wrong way of using the library? In other words, **do I need to to avoid conditionally rendering elements that have a transition, and just always render them, with the condition applied to the `in` property instead?**

This is difficult, because usually when the condition for rendering something becomes false, the data required to render it also becomes unavailable. So if I always need to render it, I end up getting a lot of errors before it has time to transition away.

But I understand that that may be unavoidable. Hope this question makes sense.

**Comment by silvenon** [comment]
It's harder to fully understand the problem without seeing some code, but I'll try. When using `TransitionGroup`, removing its children _does_ transition them out before unmounting while retaining their content even though the data is missing (I think), which sounds like what you want. Maybe a simple solution would be to wrap your transition components into `TransitionGroup`?

**Comment by jquense** [comment]
> do I need to to avoid conditionally rendering elements that have a transition, and just always render them, with the condition applied to the in property instead

Yes, when using `Transition` instead of doing `show && <MyComponent />` you should be doing `<MyComponent in={show}>`. If you want the transitioning component to mount and unmount based on `in` use the `mountOnEnter` and `unmountOnExit` props

**Comment by crswll** [comment]
I think I might be able to help explain the situation here with a very small example.

Let's say you want to fade in and out an error message based on a single property `message` which is a string.

```js
const ErrorMessage = ({ message }) => {
  const visible = !!message
  return (
    <Transition in={visible}>
      <div>{message}</div>
    </Transition>
  )
}
```

This will work fine when it's fading in, since `message` will be something that you can display. When message becomes `false` the error text disappears immediately so while the element does fade out, it's fading out an empty `div`.

I'm a Vue dev working on a React project and I find their `<transition>` component is more intuitive here since it essentially disconnects / stops updates to the element that's transitioning out as soon as their `in` condition becomes false.

Here's a super simple demo that explains it a little more : https://jsfiddle.net/crswll/ouvcw9bp/10/

Would love to hear some feedback on the correct way to approach this with React Transition Group. I can get it to work maintaining a local copy of `message` but it doesn't feel like the cleanest way.

**Comment by crswll** [comment]
@silvenon After a lot of messing around and someone smarter than me looking at it I'm finally understanding what you're talking about with TransitionGroup. It works exactly how I was imagining Transition working. I plan to to write up a PR for the documentation because it was a serious struggle for me.

**Comment by katerlouis** [comment]
It feels like this is exactly what I need, but I still don't see how TransitionGroup solves the problem here. Could somebody please explain? maybe @crswll
